Fixes #3891: Implement pre commit check to test if directives are directly referenced.

1995YogeshSharma opened the pull request:
This PR concludes #3891. 
I have limited the test to all the files in '/core'.

brianrodri committed:
Merged to upstream/develop

brianrodri committed:
Remove mysterious help directory

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

1995YogeshSharma committed:
check added for testing that all directives that do not use jinja are directly referenced

1995YogeshSharma committed:
limited test to files in /core

1995YogeshSharma commented:
@seanlip ...
oh, It shows error, `No module named bs4`

Haven't we used BeautifulSoup for anything? What should be done?

seanlip commented:
Ah, you can't use it in the scripts, I think. They run outside the oppia environment.

I suggest just doing simple string matching with the file contents (e.g. try to find "ng-template" in the file's HTML). Would that work?

prasanna08 committed:
Fix #3904: Add more rigorous tests for code classifier prediction service. (#3926)

* Add more rigorous tests for code classifier prediction service.

* Undo unnecessary changes

* Add comment

* Fix comment position.

1995YogeshSharma committed:
removed usage of bs4 module

1995YogeshSharma commented:
Oh, using bs4 made life simpler and the function more generic :stuck_out_tongue: .

I have removed its usage now, PTAL!

1995YogeshSharma commented on scripts/pre_commit_linter.py:
```diff
@@ -505,6 +511,83 @@ def _check_newline_character(all_files):
 
+def _get_ng_template_id(line):
+    """This function returns the id of the ng-template, if present.
+    The function assumes id is writted as id="id_name" i.e there is no space
```
writted --> written.

1995YogeshSharma commented on scripts/pre_commit_linter.py:
```diff
@@ -505,6 +511,83 @@ def _check_newline_character(all_files):
+    total_error_count = 0
+
+    for file_name in all_files:
+        f = open(file_name, 'rb')
```
'rb' --> 'r' 

Though this doesn't make any difference in linux, I realized that it may create problem in windows.

seanlip replied:
Also, should use `with open(..., ...) as f:`

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -124,6 +124,12 @@
     'integrations_dev/*', '*.svg', '*.png', '*.zip', '*.ico', '*.jpg',
     '*.min.js', 'assets/scripts/*', 'core/tests/data/*', '*.mp3')
 
+WHITELISTED_DIRECTIVE_IDS = [
```
The name is too generic. Perhaps something like: NG_TEMPLATE_WHITELIST or NG_TEMPLATE_DIRECTIVE_IDS_WHITELIST.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -505,6 +511,83 @@ def _check_newline_character(all_files):
+        str or None. Id of the tag or None.
+    """
+    template_pattern = re.compile(".*text\/ng-template")
+    line = template_pattern.match(line)
```
Don't redefine 'line' -- choose another name instead.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -505,6 +511,83 @@ def _check_newline_character(all_files):
+    Returns:
+        str or None. Id of the tag or None.
+    """
+    template_pattern = re.compile(".*text\/ng-template")
```
We use single quotes in python.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -505,6 +511,83 @@ def _check_newline_character(all_files):
+    line = template_pattern.match(line)
+
+    if line is not None:
+        line = line.string
```
Ditto here and below; don't redefine line.

seanlip reviewed (changes requested):
Hm, before I go any further with this review -- I think the check that's being done here is too complicated. Why can't we just check that the string "ng-template" does not exist in any HTML file, and that the string "templateUrl: UrlInterpolationService.getDirectiveTemplateUrl" or "templateUrl: UrlInterpolationService" exists in all directive JS files (or if you prefer, that "templateUrl: '" -- note the closing single quote -- does not exist)?

That way, you can piggyback on the existing checks for BAD_PATTERNS_JS_REGEXP and you can do a similar inclusion check for BAD_PATTERNS_HTML_REGEXP. I think that's more in line with the other checks and avoids detailed parsing of the files.

vojtechjelinek commented:
I agree with Sean, just check for the existence of 'ng-template' and 'templateUrl: UrlInterpolationService'.

1995YogeshSharma commented:
The check for 'ng-template' would still have to parse the whole file, so time complexity wise both are same.

Also, If I check for only that, should I make a list for NG_TEMPLATE_DIRECTIVE_HTMLS_WHITELIST and whitelist whole html files?

seanlip commented:
Yep -- my intention wasn't really to make it faster, it was more just to streamline things so that all the checks are reasonably similar and therefore extensible. We'll be adding more checks in #3905 so I think designing for extensibility makes sense here.

You could indeed make such a list, or you could incorporate it within the relevant data dict (e.g. see the 'excluded_files' property in the other dicts near the top of the file).

pandeykartikey committed:
Fix part of #2394. Added docstrings to the file core/domain/fs_domain.py (#3966)

* Fix issue #2394. Added docstrings to the file

* Made the required changes

* Made the required changes

DubeySandeep committed:
Removed multiple push of same error (#3996)

hphu committed:
fix #3954 add AdminPage page object (#3974)

* add AdminPage page object

* add reloadCollection to AdminPage

* refactor learnerDashboard

* missing semicolon

DubeySandeep committed:
Replaced the correct CSS styling for md-checkbox (#3991)

tjiang11 committed:
Add fallback to blob-builder on low-end devices (#3937)

* Add fallback to blob-builder on low-end devices

* Add no support for Auto TTS message.

* Lint.

* Throw exception to preserve stack trace.

seanlip committed:
Fix #3993: standardize check for whether autogenerated translations are shown. (#4002)

* Fix #3993: standardize check for whether autogenerated translations are shown.

* Fix import error.

Nix07 committed:
Fix part of #3950: Replace Jinja templates in the frontend with Angular (#3969)

* Fix part of #3950: Replace Jinja templates in the frontend with Angular

* Fix part of #3950: Replace Jinja templates in the frontend with Angular

* Fix part of #3950: Replace Jinja templates in the frontend with Angular

* Fix part of #3950: Replace Jinja templates in the frontend with Angular

* Fix part of #3950: Replace Jinja templates in the frontend with Angular

* Fix part of #3950: Replace Jinja templates in the frontend with Angular

* Fix part of #3950: Replace Jinja templates in the frontend with Angular

seanlip committed:
Update AUTHORS, CONTRIBUTORS, CHANGELOG and credits for v2.5.5. (#4003)

seanlip committed:
Remove integrations folders; they are now obsolete. (#4004)

* Remove integrations folders; they are now obsolete.

* Try to fix flaky Protractor tests.

1995YogeshSharma commented:
Is there a way to specify which pattern of filename to match for bad patterns?

if I am adding a check for templateUrl in BAD_PATTERNS_JS_REGEXP and want it to be applied for files only in core/. How can I go about it?

seanlip commented:
Are you able to use a glob pattern in excluded_dirs (so that you can exclude, e.g., extensions/)? That might be the simplest way, though you'll have to read the other code in that file and check that the whole thing works the way you want it to.

seanlip committed:
Add script for updating feconf. (#4006)

* Add script for updating feconf.

* Minor fix.

* Address review comments.

* Try to address another Protractor issue.

seanlip committed:
Add support for kab, ar, bn, sv languages. (#4008)

seanlip committed:
Routine update of translations. (#4009)

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Add Hungarian to list of supported languages.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix broken tests.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix translation error.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix translation error.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix errors in translations found by backend tests.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix broken tests.

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

seanlip committed:
Upgrade ui-bootstrap to 0.14.3. (#4011)

* Upgrade ui-bootstrap to 0.14.3.

* Update e2e test.

tjiang11 committed:
Allow creators to optionally disable automatic text-to-speech for an exploration. (#3918)

* Add auto_tts_enabled property to exploration; Add enable/disable button to SettingsTab

* Hide automatic text to speech options for learner.

* Lint.

* Lint.

* Add tests; suggested changes.

* Missing comma.

* Fix tests.

* Fix tests.

* Change error response type for auto_tts_enabled.

* Update golden zip file.

* Remove BrowserCheckerService

* Update golden zip schema version.

96mohitm committed:
Fix part of #2394: Added docstrings to core.controllers.reader. (#4000)

* Fix part of #2394: Added docstrings to core.controllers.reader.

* fixes

* fixes

* fixes

DubeySandeep committed:
Fix part of #3950: Removed jinja template from splash pages (#3964)

* Removed jinja template in splash/splash.html

* Removed jinja template in splash/splash_ah0.html

* Removed jinja template in splash/splash_ah1.html

* Removed jinja template in splash/splash_ah2.html

* Removed jinja template in splash/splash_ai0.html

* Removed jinja template in splash/splash_ai1.html

* Removed jinja template in splash/splash_ai2.html

* Removed jinja template in splash/splash_ai3.html

* Removed jinja template in splash/splash_at0.html

* Removed jinja template in splash/splash_at1.html

* Removed jinja template in splash/splash_at2.html

* Removed jinja template in splash/splash_av0.html

* Removed jinja template in splash/splash_av1.html

* Added Global.userIsLoggedIn in scope of splash

* Made the changes as such asked in review

* Made the changes(2) as such asked in review

* Removed identation error

* Removed identation errors(2)

* Removed newlines between buttons as asked in review

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

hphu committed:
Fix part of #3954 add LibraryPage page object (#3973)

* add LibraryPage page object

* add clickCreateActivity to LibraryPage

* adjust formatting

vojtechjelinek committed:
Fix part of #2575: RTCs refactor second milestone (#4017)

* RTCS refactor second milestone

* fix backend tests

* resolved comments

* resolved comments

* resolved comments

Anthony-Alridge committed:
Fix #3794: Complete milestone one of introducing the new fraction interaction. (#3967)

* Added tests for the object.

* Create fraction object

* Create a fraction editor with domain object and directive

* Add fraction default to list

* Add script load for FractionEditor.js

* Fix small error.

* Fix lint errors

* Fix lint errors

* Fix lint errors

* Fix lint errors

* Fix lint errors

* Address review comments

* Remove fdescribe and syntax error.

* Change property names to camelcase

* Change of schema types

* Fix linting issues.

* Address review comments

* Add object of type positive int

* Add object of type positive int

* Small style change.

* Indentation fixes

* Indentation fixes

* Address review comments.

tjiang11 committed:
Show a tooltip over hint button to suggest that the learner use a hint after a few attempts. (#4018)

* Show need a hint message after number of attempts.

* Lint.

* Update variable names.

* Don't destroy and recreate hint button on answer submissions; Have tooltip on hint button appear to the right on mobile.

* Add test for NumberAttemptsService.

* Lint.

* Extranneous comma.

* Reset number of attempts when switching to preview tab in editor mode.

seanlip committed:
Fix issue with lastComputedArrangement occasionally being undefined when calculating the destination state lists.

1995YogeshSharma committed:
Merge pull request #4031 from oppia/fix-state-layout-lag

Fix issue with lastComputedArrangement occasionally being undefined when calculating the destination state lists.

seanlip committed:
Add fallback for browsers which do not support speech synthesis. (#4028)

pranavsid98 committed:
Fix #4015: Added e2e tests for hints/solutions (#4033)

* e2e test added

* lint fix

seanlip committed:
Fix errors in production due to parameters not working correctly. (#4029)

seanlip committed:
Shorten answer group header summaries so that the existence of multiple rules can be seen. (#4030)

pranavsid98 committed:
Fix #3987: Reworked Stats tab (#3992)

* first set of stats tab changes

* change

* revamp stats tab frontend

* lint fix

* review changes addressed

* renamed handlers

* fix

* quick fix

* fixes to the front end stats

* remove console logs

* batch events

* fixes

* remove logs

* backend test fixes

* fix e2e

* few changes

* taskqueue service for analytics model update

* reworked event params

* changes

* nit fix

seanlip committed:
Text tweak for Teach page. (#4032)

vojtechjelinek committed:
Load MAthJax only where needed (#4042)

rdrsadhu committed:
fix typo in Donations Paragraph (#4040)

Add missing spaces in I18N_CONTACT_PAGE_PARAGRAPH_11

vojtechjelinek committed:
Added iamge load into configuration to get rid of errors (#4041)

ishucr7 committed:
Replace Jinja templates in frontend with Angular (#4037)

fix part of https://github.com/oppia/oppia/issues/3950

ishucr7 committed:
Fix part of #3950: Replace Jinja templates in frontend with Angular (#4019)

* Replace Jinja templates in frontend with Angular
fix part of https://github.com/oppia/oppia/issues/3950

* Updated preferences.html

* Updated preferences.html .

* Updated preferences.html

ishucr7 committed:
Fix part of #3950: Replace Jinja template in frontend with Angular (#4050)

fix part of: https://github.com/oppia/oppia/issues/3950

seanlip commented:
Hi @1995YogeshSharma, friendly ping? This PR has been sitting for a while.

tjiang11 committed:
Only record stats for solutions if not in preview mode. (#4049)

tjiang11 committed:
Fix #3976: Show supplemental interactions in embedded explorations (#4052)

* Show supplemental card in embedded explorations.

* Show supplemental interaction in embed.

* Increase min-height for supplemental interactions.

* Change margin-bottom to include footer in all screen sizes.

DubeySandeep committed:
Fix #4046:  SITE_NAME replacement error fixed in Licence explanation page (#4051)

* SITE_NAME replacement error fixed in Licence explanation page

* Removed Merge conflict from Base.html

DubeySandeep committed:
Added proper css for CodeRepl preview (#4053)

seanlip committed:
Reformat frontend error message so that StackDriver hopefully catches it. (#4045)

tjiang11 committed:
Fix #3791: Paragraph spacing fix (#4048)

* Fix paragraph spacing.

* Combine word-wrap CSS. Apply conversation-skin-learner-answer-content at a more local level to more closely match other uses.

* Extranneous line.

* Minor update in conversation_skin_embed_directive to match CSS.

pranavsid98 committed:
Fix #3958: One-off job to use existing stats data to create new ExplorationStatsModels (#3959)

* create job for migration

* tests

* post merge follow up

* stats audit job

* lint fixes

* small functionality change

* addressed comments

* changes

* exp reverts

* changes

* added get multi

* changes addressed

* fixes

* lint fix

* completion counts

* changes

* addressed minor review comments

* changes

* changes addressed

* changes

* changes

* final nit changes

vibhor98 committed:
Fix part of #2394: Add docstrings to core.domain.subscription_services.py (#4054)

* Remove whitespace

* Add docstrings

* Add requested changes

* Remove linting errors

* Add requested changes.

seanlip committed:
Remove dead code. (#4064)

pranavsid98 committed:
Fix errors in the one-off job for stats migration (#4061)

* handle errors

* changes

* addressed review changes

* fix issue 1

* nit changes

* changes

* changes

* add job for identifying conversion errors

* review changes

* fixes

ishucr7 committed:
Fixed  Adjust the tooltip displayed when a user hovers over the globe icon on the library page (#4043)

fix https://github.com/oppia/oppia/issues/2943

seanlip committed:
Simplify karma coverage specification. (#4066)

BenHenning committed:
Fixed #3288: Add tests for stats_visualizations_info (#4035)

* Cleanup get_calc_output in stats_services and stats_continuous (the
former is now private and the latter is no longer used publicly, so it
was removed).

* Add tests for get_visualizations_info

* Add tests for get_visualizations_info. Include interaction_id in
StateAnswersCalcOutputModel and fix a bug where visualization info can
be retrieved for old interaction IDs due to a stale calculation ID.

* Fix broken stats_domain test.

* Address review comments.

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

prasanna08 committed:
Implement state id mapping model. (#3963)

* Add state id mapping model.

* Address review comments.

* Addressed review comments.

* Fix backend tests

* Fix backend tests

* Fixed backend tests and addressed review comments

* Fix end to end tests.

* Addressed review comments.

* Addressed review comments.

Anthony-Alridge committed:
Fix part of #3794: Complete milestone 2 of fraction interaction (#4067)

* Added tests for the object.

* Create fraction object

* Create a fraction editor with domain object and directive

* Add fraction default to list

* Add script load for FractionEditor.js

* Fix small error.

* Fix lint errors

* Fix lint errors

* Fix lint errors

* Fix lint errors

* Fix lint errors

* Address review comments

* Remove fdescribe and syntax error.

* Change property names to camelcase

* Change of schema types

* Fix linting issues.

* Address review comments

* Add object of type positive int

* Add object of type positive int

* Small style change.

* Indentation fixes

* Indentation fixes

* Address review comments.

* Create equivalence rule.

* Complete fraction rules service.

* Create validation service from the interaction.

* Small style fixes.

* Small style fixes.

* Small style fixes.

* Change name offraction entry in rules templates to fractionInput

* Address review comments.

* Adjust indentation for multi line comment.

Commit:
.gitignore

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

DubeySandeep committed:
Fixed #3146 : Reset Button functionality in preview mode (#4078)

DubeySandeep committed:
Fix #4001: Learner Dashboard Unpublished exploration issue fixed (#4039)

* Made Deleted and Unublished activity id as Nonexitent for learner

* Made Deleted and Unublished activity id as Nonexitent for learner(2)

* Made changes to the test files

* Made changes to the i18n files

* Made changes to i18n, arranged translation in AlphabeticOrder

* Made the changes which were asked by the reviewer

* Made the changes which were asked by the reviewer.

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

joefedota committed:
Fix #4055: Replaced redundant CSS selectors. (#4059)

Fix part of #4055: Replaced redundant CSS selectors.

shubha1593 committed:
Fix part of #3826: Extract services from app.js (#4079)

* Add docstrings to core.domain.user_jobs_continuous.py

* Extract DeviceInfoService from app.js

* Extract WindowDimensionsService and UrlService from app.js

* Replace window by $window, so possible to mock it for unit test.

* Remove extra line.

* Remove currentLocationService from app.js

bching committed:
Fix part of #3826: extract from EditorServices.js (#3970)

* Partial Fix #3826: extract from EditorServices.js

* explorationData to ExplorationDataService
* editorContextService to EditorStateService
* angularNameService to AngularNameService

* fix merge mistakes

* fix e2e errors and linting errors

* remove new line linter error

* Update StatisticsTab.js

fix line length

* fix format and remove includes

* Revert "Merge remote-tracking branch 'origin/extract-editorservice' into extract-editorservice"

This reverts commit d98998ce95be90df04dfe1e8816245a569c6d08f, reversing
changes made to 84dcff246f3fd23e9979db1e78230ca93d7fee93.

* move to previous line

* add periods

* remove erroneous additions

* remove line

seanlip committed:
Upgrade SSL library version; remove obsolete challenge-response handler. (#4081)

tjiang11 committed:
Stop cards from going offscreen in mobile learner view. (#4084)

tjiang11 committed:
Rename CSS property oppia-learner-continue-button to oppia-learner-confirm-button; Make the Submit button in ItemSelectionInput the same style as the Continue button; Move informational text in ItemSelectionInput from bottom-right to top of the interaction so that it is more visible; Add extra text to ItemSelectionInput indicating to the learner that they may select more than one choice. (#4083)

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

Nix07 committed:
Fix part of #3954: Created a Page Object for Thanks Page (#4085)

* Fix part of #3954: Created a Page Object for Thanks Page

* Fix part of #3954: Modified Page Object

* Fix part of #3954: Modified Page Object

* Fix part of #3954: Simplified Page Object

* Fix part of #3954: Corrected spacing

seanlip committed:
Routine update of translations. (#4082)

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Add Hungarian to list of supported languages.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix broken tests.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix translation error.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix translation error.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix errors in translations found by backend tests.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Fix broken tests.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

* Localisation updates from https://translatewiki.net.

anmolshkl commented:
Hi @1995YogeshSharma, ping!

prasanna08 committed:
Implement one off MapRreduce job to generate state id mapping model for explorations. (#4088)

* Implement MR job for state id mapping model.

* Addressed review comments

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into develop

brianrodri committed:
Add column for Addressed? in text input.

tjiang11 committed:
Prevent hint tooltip from going offscreen on mobile (#4087)

* Make tooltip pop right on hint button in supplemental interactions on mobile; Keep hint button visible after exhausting hints.

* Lint.

* Remove two-way binding

tjiang11 committed:
Update Authors, Contributors, Changelog, Credits for Release 2.5.6. (#4095)

tjiang11 committed:
Minor updates to CHANGELOG v2.5.6 after proofreading (#4097)

sriordan9 committed:
Added underline to 'Back to collection' link (#4096)

prasanna08 committed:
Add state id mapping mapreduce job to the job registry (#4099)

pranavsid98 committed:
Fix error in one off stats migration job. (#4068)

* make defaultdict

* lint fix

* try-catch

* added unicode keys

seanlip committed:
Fix exploration graph directive URL. (#4098)

vojtechjelinek committed:
Renamed RTC_specs to RTC_definitions and removed useless __init__.py file. (#4101)

* Renamed specs to definitions and remove useless __init__.py file

* renamed constant

BenHenning committed:
Fix #3289: Introduce domain objects for answer calculation output (#4036)

seanlip committed:
Remove feedback button from mobile learner view. (#4093)

brianrodri committed:
Fix 3946: Add visualization for Item selection inputs (#4080)

* Merged to upstream/develop

* Remove mysterious help directory

* .gitignore

* Give ItemSelectionInput the same answer visualization specs as TextInput.

* Initial specification of EnumeratedFrequencyTable

* Add basic functionality for new EnumeratedFrequencyTable visualization.

* Use singlequotes for strings

* Revert .gitignore

* Improve ID usages.

* Remove debugging helper 'escaped_answer'

* Improve naming of answer toggle function.

* Use answer in ID instead of table

* Better naming in ng-repeat

* Simplify key generation

* Better arg naming

* Remove padded blank line

* Format improvements to ItemSelectionInput.py

* Variable name changes and merge keying data into a map call.

* Use camelCase variable names

* Move all styling to css code.

* Better display name

* Use ternary if expression

* Use  instead of adding attributes to .data

* Use $index and ng-show for visiblity control.

* Improve documentation of EnumeratedFrequencyTable.

* More doc improvements

* better variable naming

* Avoid .fill

* Use angular-html-bind

* Show first element by default.

* Update docs to show that first entry is shown by default.

* Whoops... always use === :^)

* Minor wording fixes

* Style answer-rank links through class.

* Give item tables their own class and better top-margin.

* Remove extraneous space.

* Add odd-coloring to answer tables. Fix docstring and missing to_raw_type call.

* Update tests to not call to_raw_data directly.

shubha1593 committed:
Fix part of #3825: Rename angular services in Group 5 (#4111)

* Add docstrings to core.domain.user_jobs_continuous.py

* Rename alertsService to AlertsService.

* Rename autoplayedVideosService to AutoplayedVideosService.

* Rename explorationContextService to ExplorationContextService.

* Rename oppiaExplorationHtmlFormatterService to ExplorationHtmlFormatterService.

* Remove extra line.

brianrodri committed:
Refactor extensions.answer_summarizers.models logic (#4092)

* Merged to upstream/develop

* Remove mysterious help directory

* .gitignore

* Introduce _HashableAnswer class to make _count_answers more intuitive.

* Doc string improvements.

* Make doc string format consistent across file.

* Simplify comment.

* Move num_results capping to _count_answers.

* Allow n to be None.

* Add returns doc string.

* Expand logic for more space to add useful comments.

* Make _get_hashable_value a static method of _HashableAnswer

* Rename to _make_hashable

* Minor spacing changes

* Make _to_hashable a class method so it can be used recursively.

* Fix lint errors

* Disable protected access, accessing member of same class

* Revert personal file changes.

* Add parenthesis to docs for readability

* Rename to _hashable

* rename _count_answers to _get_top_answers

* Consistent spacing between class doc strings and methods

* Use better spacing, remove duplicate logic

* Rename to _HashedAnswer.

* Remove assumption that 1-element lists equal 1-element.

* Rename into _get_hashable_value.

* Keep as class method.

* Better name for class member values.

* Documentation improvements.

* Remove wrong doc statement.

* typo: %s/_answer_hash/answer_hash

* Didn't split into smaller pieces... Lots of changes here:

* Fix indentation

* Rename to _HashedAnswerDict

* Remove unneccesary pylint disable

* Rename member to answer_dict

* Fix typo in original code

* Better wording.

* Better wording.

* Use simpler doc

* Fix incorrect logic of FrequencyCommonlySubmittedElements calculation

* Fix incorrect logic of FrequencyCommonlySubmittedElements calculation

* Different spacing for list comprehension

* More simplifications.

* Allow ElementCounting to count arbitrarily complex elements.

* Fix aggregation of hashed values

* Doc changes

* Add missing parenthesis

* Use operator.itemgetter

* Doc and spacing changes

* Doc and spacing changes

* Add comma for readability

* Doc and spacing changes

* Doc and spacing changes

* Better doc string

* Rename to _get_top_answers_by_frequency

* Consistent collection comprehension spacing.

* Remove one-too-many loops

* Make comprehension var names more consistent.

* Prefer update to manually adding frequencies

* Rename constant.

* Avoid using the term collection.

* Stop using "key" in _HashedValue.

* Add missing parenthesis.

* Replace the term element with individual answer.

* Make CLASSIFICATION_CATEGORIES a set

* Use a frozenset.

* Remove checks for set, JSON values can't have them.

* Reorganize comments.

* Update docstrings to match new behavior.

* Remove some lines.

* Better spacing.

* Use itertools to express intent clearly.

* More consistent docstrings

* Add kwarg names to help self-document.

* Simplify sum

* Move HashedValue to utils.

* Add helpers to utils

* Add test coverage for BaseCalculation.

* More test coverage.

* Test coverage for Categorization.

* Add kwarg-naming to test call.

* Use more commas.

* Add comments to help separate answer dict types.

* Better test name.

* Add 2-line spacing between classes.

* Add tests for FrequencyCommonElements.

* Fix line-too-long.

* Fix dangerous-default-value.

* Fix unused-import.

* Give inf_args a better name.

* Better division of calculation id fetching.

* Better naming for infinite args.

* Fix import ordering, better variable naming.

* More complete doc string for _create_answer_dicts_list.

* Better naming for dummy variables.

* Add pass to empty class.

* Make classify_categories doc consistent.

* Use lists for args of _create_answer_lists.

* Use None as default value classify_categories.

* Fix superfluous-parens and unused-import lint errors

* Rename as to_pair

* Make answer_counter public.

* Rename Answer to HashableAnswer.

* Change tests to use new HashableAnswer.

* Use unique objects to test validation.

* Update to use HashableAnswer.

* Better initialization for frequency list.

* Require answer_occurrences.

* Don't get rid of answer_occurrences arg entirely.

* Force passing empty list.

* Add truncation for FrequencyCommonlySubmittedElements.

* Simplify helper functions.

* Move helper constants to correct class.

* Simplify simple answer list.

* Simplify return expression.

* Simplify args creation.

* Fix endless loop.

* Add unit test for get_hashable_value.

* Access get_hashable_value through module.

* Use set comprehension.

* Update docstring.

* Simplify test.

* More simplification.

* Get rid of custom times_spent/session_ids.

* Get rid of useless class constants.

* Better dummy variable.

* Simplify calculation_id check.

* Add helpful comments.

* Make answer_list mandatory.

* Add TODO for msl@ and I.

* Remove create_answer_dicts entirely, replace with list comprehensions.

* Keep test structure consistent.

* Keep things consistent.

* Clarify docstring.

* Remove unused import.

* Move away from polluting stats_domain with logic.

* Change helper methods to clean-up calls.

* Give clearer name.

* Change argument name.

* Better dummy variable names.

* Use generator expressions instead of function call.

* Use expression instead of map.

* Remove extra line.

* Add docstring for hashable answer.

* Add line for consistency.

* extra p

* Better doc strings.

* Add coverage for 10< total individual answers.

* Move doc string.

* Make calculation more consistent.

* Better dummy variable names.

* Update core.jobs to use utils.OrderedCounter.

* Remove extra utils import.

* Remove unused collections import.

* Save this change for a different PR.

* Remove add_answer API from AnswerFrequencyList.

* Change docstring to highlight intent as an output.

* Change expected-output to respect the order elemens are added in.

* Pass a list to the AnswerFrequencyList constructor, not a generator.

* Remove debug message.

* Add comment to explain change.

* Add Args section to doc string.

* Add docstring.

* Clarify doc-strings.

* Use nicer iterable(*) syntax.

* Plural numbers.

* *Correct* plural.

* Change expected output to respect submission ordering.

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into addressed-column

brianrodri committed:
Use upstream .gitignore

brianrodri committed:
Create template for showing addressed column.

seanlip committed:
Update PULL_REQUEST_TEMPLATE.md (#4117)

prasanna08 committed:
Fix ExplorationStateIdMapping MapReduce job. (#4113)

* Fix ExplorationStateIdMapping MapReduce job.

* Addressed review comments.

vojtechjelinek committed:
Fix part of #2575: Interactions refactor first milestone (#4115)

* moved js and html to /directive folder

* Moved directive HTML to separate files

* Moved HTML files to root of interaction and styles to separate css files

* Added UrlInterpolationService to directives

* Fixed errors and move objects_defautls

* Fixed frontend tests

* Fixed backend tests

* Fixed linter issues

* Fixed linter issues

* Fixed linter issues

* added tests for utils and romoved useless code from base_tests

brianrodri committed:
temp

pranavsid98 committed:
Fix error in GenerateV1StatisticsJob (#4120)

* changes

* fixes

* fixes

* fixes

* fixes

nithusha21 committed:
Fix part of #3826: Extract ExplorationWarningsService into induvidual file. (#4106)

* extracted ExplorationWarningsService into a seperate js file

* fixed linting error

* added new js file to html pages

* removed extra newlines

* removed extra import

brianrodri committed:
Fix lint issues.

prasanna08 committed:
Log error when ExplorationCommitLogEntry model instance is not found for (#4126)

a particular exploration version.

jervis446 committed:
Fix part of #3826: Extract services from i18n.js  (#4118)

* modified files for necessary update

* i18n.js to individual files

* remove linting error

* add npm log

* add npm log

* npm update

* update npm

* update npm

* update npm log

* delete npm log

* remove linting error

* remove linting error

* remove linting error

* remove js Linting error

* Update files

* update changes

* add spaces

* Update changes

* update to check travis test

* final check

* remove space

aks681 committed:
Fixed part of #3826 - editorFirstTimeEventsService (#4124)

* fixed part of #3826 - editorFirstTimeEventsService

* fixed part of #3826 - editorFirstTimeEventsService

* fixed part of #3826 - editorFirstTimeEventsService

* fixed part of #3950 in creator_dashboard.html

* #3826 removed script line in collection_editor.html

* moved up last comment line

* reduced spacing in comments

prasanna08 committed:
Scroll down to latest Oppia response in learner card when in mobile view (#4110)

* Autoscroll to Hint button in mobile view.

* Auto scroll to new feedback in learner card.

* cleanup!

* Addressed review comments.

prasanna08 committed:
Fix ExplorationStateIdMapping job: Yield errors rather than raising Exception in ExplorationStateIdMapping (#4128)

* Yield errors rather tha raising Exception in ExplorationStateIdMapping
job.

* Fetch commit cmds from snapshot metadata model

Anthony-Alridge committed:
Fix remainder of #3794: Add a fraction input interaction (#4100)

* Added tests for the object.

* Create fraction object

* Create a fraction editor with domain object and directive

* Add fraction default to list

* Add script load for FractionEditor.js

* Fix small error.

* Fix lint errors

* Fix lint errors

* Fix lint errors

* Fix lint errors

* Fix lint errors

* Address review comments

* Remove fdescribe and syntax error.

* Change property names to camelcase

* Change of schema types

* Fix linting issues.

* Address review comments

* Add object of type positive int

* Add object of type positive int

* Small style change.

* Indentation fixes

* Indentation fixes

* Address review comments.

* Create equivalence rule.

* Complete fraction rules service.

* Create validation service from the interaction.

* Small style fixes.

* Small style fixes.

* Small style fixes.

* Change name offraction entry in rules templates to fractionInput

* Address review comments.

* Add validation check for equals rule that is not in simplest form.

* Adjust indentation for multi line comment.

* Create html and js directives for learner facing portion of fraction interaction

* Create new rule checking if fraction has a fractional part.

* minor style changes.

* minor style changes.

* minor style changes.

* Add relevant fraction files.

* Minor style changes.

* Minor style changes.

* Edit specs causing tests to fail.

* Indentation fixes.

* Indentation fixes.

* Fix failing backend test

* Changing the names of rules.

* Style fixes.

* Improve user input error messages.

* Add better quality fraction input thumbnail.

* Add gap between form and error message.

* Add validation check for non integers in rule definitions

* Fix hanging UI bug when non integers are entered

* fix tests to conform to correct interface

* Add filter test for isInteger in form builder

* Add filter test for isInteger in form builder

* Style checks

* Add input validation whilst user is entering input.

* Refactor files to match new file layout for extensions

* Refactor files to match new file layout for extensions

* Address review comments.

* Remove unused name attr in fraaction editor

nithusha21 committed:
Fixed part of #3950: Replace Jinja templates in feedback_tab.html with Angular (#4123)

* replaced jinja if statements with angular ng-if

* removed extra newline

* changed variable name

* fixed indentation

* fixed indentation

BenHenning committed:
Remove globe from exploration summary cards. (#4132)

Nix07 committed:
Fix #4102: Auto Hide of Navbar (#4108)

* Fix of #4102: Auto Hide of Navbar

* Fix of #4102: Fixed Indentation

* Added NavBar Hiding Functionality

* Removed auto hide on desktop

* Fixed Minor Typo

* Added headroom to manifest.json

* Fix #4102: Hiding the navbar automatically when user scrolls down

prasanna08 committed:
Check for empty change_list before proceeding. (#4136)

seanlip committed:
Move isInteger polyfill to app.js. (#4134)

bching committed:
Fix #4090: limit number of progress dots for narrow window (#4129)

* Limit number of dots to 6 if window is narrow

* add parentheses

pranavsid98 committed:
Fix bugs in GenerateV1StatisticsJob (#4125)

* add fixes

* more fixes

* added logs

* redefined vars

* small fix

* small fix

* fixes according to job run

* final fixes

* change phrasing

* lint fixes

* quick fixes

* fix errors

* blank line removal

* remove app yaml changes

* changes

* fixes

* fix audit job

* final fixes

* handling END states

* minor fix

* fixes

* changed logs

* new logs

* fixes

seanlip committed:
Fix #4140: correctly generalize permissions for creating collections. (#4142)

* Fix #4140: correctly generalize functionality for creating collections.

* Fix e2e test.

akumar1503 committed:
Fix part of #3826: Extract stateEditorTutorialFirstTimeService into induvidual file (#4137)

* replaced all stateEditorTutorialFirstTimeService with StateEditorTutorialFirstTimeService and created a new file StateEditorTutorialFirstTimeService.js

* added StateEditorTutorialFirstTimeService.js as script in exploration_editor.html

brianrodri committed:
Merge remote-tracking branch 'upstream/develop' into addressed-column

brianrodri committed:
Call solution verification service

brianrodri committed:
Pass explorationId and stateName via attrs.

brianrodri committed:
Make addressed column update

brianrodri committed:
Give addressed column to all eligible interactions.

brianrodri committed:
Return queue.yaml rates to normal

brianrodri committed:
Remove unused dependency.

brianrodri committed:
Simplify some code and comments.

brianrodri committed:
Make answer column wider.

brianrodri committed:
Use ng-if for showing addressed cells.

brianrodri committed:
Functions statements always start on newline.

brianrodri committed:
Use explorationStatesServices in visualizations.js

brianrodri committed:
Move all addressed stuff to async work in StatisticsTab.js.

brianrodri committed:
Fix lint issues.

brianrodri committed:
Fix indentation.

brianrodri committed:
Fixed indentation.

brianrodri committed:
Make verifySolution sync.

brianrodri committed:
Update StateSolution to use sync verifySolution.

brianrodri committed:
Update SolutionEditorDirective to use sync verifySolution.

brianrodri committed:
Update StatisticsTab to use sync verifySolution.

brianrodri committed:
Update Spec to use sync version of verifySolution.

brianrodri committed:
Update ResponsesServices to use sync verifySolution.

brianrodri committed:
Better naming.

brianrodri committed:
Give better name to isValid.

brianrodri committed:
Simplify if-else branching.

brianrodri committed:
Simplify if-else logic for ResponsesService.

brianrodri committed:
Simplify StateSolution if-else logic.

brianrodri committed:
Remove $q.

brianrodri committed:
Split return and build.

brianrodri committed:
Fix max-len issue.

brianrodri committed:
Fix comment.

brianrodri committed:
Comments on their own line.

brianrodri committed:
Minor wording.

brianrodri committed:
Update core.domain.stats_services_test.

BenHenning committed:
Fix #4076: Update end exploration (within collection) headers to better illustrate that the user should return to the collection (#4130)

* Add new example explorations & update demo collection to ensure that the
collection can be fully completed for testing purposes.

* Update end of exploration content based on whether you're in a
collection and to match the provided mocks.

* Remove Go to Library button & make 'back to collection' text consistent
with return to library text.

brianrodri committed:
Add missing column header.

brianrodri committed:
Update visualizations models test.

brianrodri committed:
Merge pull request #4147 from brianrodri/addressed-column

Fix #3906: Show Addressed column in State Stats Modal

pranavsid98 committed:
UI fixes to the stats tab (#4150)

* ui fixes

* switch flags

* fixes

* fixes

aks681 committed:
Fixed part of #3950 in creator_dashboard.html (#4127)

* fixed part of #3826 - editorFirstTimeEventsService

* fixed part of #3826 - editorFirstTimeEventsService

* fixed part of #3826 - editorFirstTimeEventsService

* fixed part of #3950 in creator_dashboard.html

* #3826 removed script line in collection_editor.html

* removed extra divs

* made the review changes

aks681 committed:
fix part of #3954  - LearnerDashboardPage.js (#4138)

* fix part of #3954 - Learner Dashboard Page

* fixed lint errors

* removed LEARNER_DASHBOARD_URL from general.js

* fixes to learnerDashboard.js

* some fixes and more functions added

* made the review changes

* changed page object names

* added subscription and feedback sections

* made the review changes

seanlip committed:
Fix #4152: fix bug in solution verification service. (#4154)

BenHenning committed:
Update deploy script (#4156)

* Update deploy script to only open the library & logging pages for a test deployment when the correct version is serving on app engine.

1995YogeshSharma commented:
I have looked and made changes, I'll push the PR after checking locally (I am having problem with excluding the js files in /extensions from linting)

prasanna08 committed:
Catch Exception when failing to retrieve explorations. (#4157)

lunareve committed:
Fix part of #3826: Extract ExplorationRightsService from EditorServices.js. (#4139)

* Extract ExplorationRightsService from EditorServices.js.

* Remove indent

seanlip committed:
Fix #4159: update link to coding style guide (#4160)

seanlip committed:
Fix #4164: error with running ExplorationStateIdMappingJob (#4165)

anmolshkl committed:
Add front-end prediction service for TextInput interaction. (#3861)

Add TextInputPredictionService for predicting answer groups for a TextInput.

brianrodri committed:
Improve readability of extensions.answer_summarizers.models_test (#4171)

* Use upstream gitignore

* Keep cause and effect for tests closeby.

* Whoops

* Keep cause and effect nearby.

* Rename variable to split_len.

* Add extra spaces for consistent formatting.

aks681 committed:
fix part of #3954 - ExplorationPlayerPage.js (#4170)

* refactored some objects to explorationplayerpage

* minor fixes

* fixed lint errors

* fixed lint errors

* reduced line sizes

* changed all tests to use ExplorationPlayerPage.js

* made review changes

* made review changes

pranavsid98 committed:
Fix #4169: Fixes Key error due to plus sign replacements (#4172)

1995YogeshSharma committed:
Merge remote-tracking branch 'origin/issue-3891-final' into issue-3891-final

1995YogeshSharma committed:
shifted the direct referencing of directives check to bad patterns check.

1995YogeshSharma commented:
sorry, a lot of random commits have cropped in. I am closing this PR and opening a new one.
